components/cloud/aws/src/s3.rs

```diff
@@ -198,20 +221,63 @@ impl S3Storage {
         Ok(S3Storage { config, client })
     }
 
+    fn maybe_assume_role<P, D>(
+        config: Config,
+        cred_provider: P,
+        dispatcher: D,
+    ) -> io::Result<S3Storage>
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        if let Some(role_arn) = config.role_arn.clone() {
+            // try use role arn anyway with current creds when it's not nil.
+            let bucket_region = none_to_empty(config.bucket.region.clone());
+            let bucket_endpoint = config.bucket.endpoint.clone();
+            let region = util::get_region(&bucket_region, &none_to_empty(bucket_endpoint))?;
+            // cannot use the same dispatcher because of move, so use another http client.
+            let sts = StsClient::new_with(util::new_http_client()?, cred_provider, region);
+            let duration_since_epoch = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap();
+            let timestamp_secs = duration_since_epoch.as_secs();
+            let cred_provider = StsAssumeRoleSessionCredentialsProvider::new(
+                sts,
+                none_to_empty(Some(role_arn)),
```

Comment:
...here, so we can probably do:

```suggestion
                String::clone(config.role_arn.as_deref().unwrap()),
```

Since `StringNonEmpty` can be derefed to `String`.